### ISSUE TITLE
DataMatrix not recognized with PyPi python package

### ISSUE THREAD

**Issue by guylevin** [zxing-cpp/zxing-cpp#695]
I'm trying to read a small Data Matrix in pretty difficult photo taken in a bit of an angle and from pretty far distance

When testing it with the online wasm example reader provided in the readme, it seems to work
https://zxing-cpp.github.io/zxing-cpp/demo_reader.html

But with the python wrapper it does not

```
  img = cv2.imread("assets/image.jpg")
  
  results = zxingcpp.read_barcodes(img)

  for result in results:
      print(
          "Found barcode:"
          f'\n Text:    "{result.text}"'
          f"\n Format:   {result.format}"
          f"\n Content:  {result.content_type}"
          f"\n Position: {result.position}"
          f"\n Orientation: {result.orientation}"
      )
```

I noticed that the python wrapper does not support the try_harder option, so I also tried it with the zxing python library which does seem to support it, but it didn't work either

```
    reader = zxing.BarCodeReader()

    img = cv2.imread("assets/img.jpg")

    img_pil = Image.fromarray(img)
    result = reader.decode(img_pil, try_harder=True)

    print("result", result.raw)
```

**Both of them work when I crop the image**
**Both of them work with a higher resolution image**

Unfortunately, I can't publish a sample photo here, but I hope there are suggestions on how to overcome this issue

This is the barcode after cropping
![image](https://github.com/zxing-cpp/zxing-cpp/assets/4077233/75508908-2c5d-43e9-b842-fc78cfbb7c7d)

**Comment by axxel** [comment]
`try_harder` is always enabled in the python wrapper. I would expect that the wasm wrapper detects exactly the same images as the python wrapper does. Are you sure you are using the latest version of the python wrapper? Did you install a precompiled wheel from PyPi or did you compile the module yourself?

Please try to provide a sample that shows the problem, so I can reproduce it. If the full image contains sensitive data, maybe you can simply erase the sensitive parts while still preserving the property of the DataMatrix code not being detected?

**Comment by kimstp** [comment]
I am having the same issue with the python wrapper on this [image](https://specify-attachments.science.ku.dk/fileget?coll=NHMD+Vascular+Plants&type=O&filename=sp68879436369884925558.att.jpg&downloadname=NHMD-681096.jpg&token=b372879b318d68d2f93369b0bcfe56b4%3A1704396347).

The datamatrix code is found using the wasm example, but fails to be found using the python wrapper.
I am using the latest pypi package zxing-cpp version 2.2.0.

A simplified version of my Python code looks like this
```python 
import cv2
import zxingcpp

img = cv2.imread("myimage.jpg")
dmcodes = zxingcpp.read_barcodes(img, formats=zxingcpp.DataMatrix)
print(dmcodes)
```

Hope this gives you a test case to work with.

**Comment by axxel** [comment]
Thanks @kimstp for the sample. I was able to find the problem, which is that those pre-build python packages (at least the macOS and the Linux ones) on PyPi are in fact compiled with an (old) c++17 compiler (details can be found [here](https://github.com/pypa/cibuildwheel/discussions/984#discussioncomment-1944937)). Therefore, the position independent DataMatrix detection does not work. 

For Linux x86_64 wheels there is hope (see [here](https://github.com/pypa/manylinux)). For macOS, I don't know how to interpret the situation regarding minimal macOS versions.

The python package build on my local machine works just fine. So for the time being, I suggest you build the wheel yourself from source with the following command (you need to `pip uninstall zxing-cpp` first):

```sh
python -m pip install zxing-cpp==2.2.0 --no-binary zxing-cpp
```

**Comment by axxel** [comment]
As mentioned before, for Linux x86_64 wheels I found a solution (see above commit). You can find a binary wheel inside this [artifacts.zip](https://github.com/zxing-cpp/zxing-cpp/actions/runs/7421538247/artifacts/1149841372). The PyPi published packages will only be updated on the next release.

If someone on the macOS side could investigate how to improve the situation there and report back, that would be much appreciated.

**Comment by kimstp** [comment]
Thanks @axxel - installing from source on MacOS 13.6.3 as you suggested works fine. I tried to use the binary wheel for python 3.10 from your artifacts.zip file but it does not work on my installation.

So to summarize, building from source works on MacOS 13.6.3 with Xcode ver. 15.1 which seems to support C++20 as stated by [Apple](https://developer.apple.com/xcode/cpp/). As can be seen from this language feature support table, it seems that you need  at least Xcode 15 to maximize the number of supported c++20 features. The corresponding Apple clang compiler is version 15.0.0.

**Comment by axxel** [comment]
@kimstp Thanks for confirming that the source-build solution works for you. I'm puzzled by your statement that the binary wheel from the artifacts "does not work". You mean it does not detect the code from the sample (that would be expected, because I only changed the Linux build) or you mean it does not work at all (like not installable)?

**Comment by guylevin** [comment]
> Thanks @kimstp for the sample. I was able to find the problem, which is that those pre-build python packages (at least the macOS and the Linux ones) on PyPi are in fact compiled with an (old) c++17 compiler (details can be found [here](https://github.com/pypa/cibuildwheel/discussions/984#discussioncomment-1944937)). Therefore, the position independent DataMatrix detection does not work.
> 
> For Linux x86_64 wheels there is hope (see [here](https://github.com/pypa/manylinux)). For macOS, I don't know how to interpret the situation regarding minimal macOS versions.
> 
> The python package build on my local machine works just fine. So for the time being, I suggest you build the wheel yourself from source with the following command (you need to `pip uninstall zxing-cpp` first):
> 
> ```shell
> python -m pip install zxing-cpp==2.2.0 --no-binary zxing-cpp
> ```

Thanks @axxel 
It works great on Linux after building the wheel locally like you suggested.

**Comment by kimstp** [comment]
Sorry @axxel for being imprecise in my statement about the wheel binary in the artifacts.zip file.
I cannot import the file in python and cannot tell you if it detect the data matrix code or not. It seems that the python 3.10 version of the wheel is build with python 3.12. I did not investigate further if this holds for the other wheel packages.

**Comment by axxel** [comment]
> I cannot import the file in python

That is worrisome. The thing is, I did not change anything regarding the construction of the mac os python wheels since the 2.2.0 release. So if the wheels on pypi.org work for macos, so should the ones inside the linked artifacts.zip. Can you double check that the wheel from [pypi.org](https://pypi.org/project/zxing-cpp/#files) works when manually downloaded and installed while the one from the artifacts.zip with the same name does not?

**Comment by guylevin** [comment]
I'm not sure if this is related, but I've just stumbled on another problematic reading.
This time it's QR and not DataMatrix, trying to read it unsuccessfully with the locally built wheel suggested here.
Google Lens and Microsoft Bing are able to read the code from the same images.

I even used an object detection model to crop the barcode with different "buffer area" around it, but it still doesn't read it.
The image is a bit blurry.

![detection_1 (2)](https://github.com/zxing-cpp/zxing-cpp/assets/4077233/ccdc5379-aa4f-4c80-a29f-50a1612cc5b9)
![detection_1 (1)](https://github.com/zxing-cpp/zxing-cpp/assets/4077233/dceb8c06-a823-4184-a0c0-839ca4a1c881)
![detection_1](https://github.com/zxing-cpp/zxing-cpp/assets/4077233/27678b82-65c6-4546-99cf-c60659635fa2)

**Comment by axxel** [comment]
> I'm not sure if this is related

This is unrelated. The problem is caused by the noise/low-res/compression artifacts. Instead of having 'black' and 'white' there is a substantial portion of the symbol 'greyish'.

> Should the library do it automatically?

No. This is trivial to do outside and I don't want to unnecessarily increase the API.

**Comment by axxel** [comment]
@kimstp May I kindly ask whether you had the chance to test the wheel in the artifacts vs the one from pypi.org (see my ealier comment: https://github.com/zxing-cpp/zxing-cpp/issues/695#issuecomment-1880085157)?

**Comment by kimstp** [comment]
Sorry about the silence @axxel. I just took the wheel files from your artifacts.zip file and attempted to install and the installation is fine:
```sh
(venv) kimstp@hodor NHMDlabelreader % pip install ~/Downloads/artifact/zxing_cpp-2.2.0-cp310-cp310-macosx_10_9_universal2.whl 
Processing /Users/kimstp/Downloads/artifact/zxing_cpp-2.2.0-cp310-cp310-macosx_10_9_universal2.whl
Installing collected packages: zxing-cpp
Successfully installed zxing-cpp-2.2.0
```
But when I attempt to import the module in python I get a ModuleNotFoundError:
```python
(venv) kimstp@hodor NHMDlabelreader % python
Python 3.10.13 (main, Aug 25 2023, 01:44:57) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import zxingcpp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'zxingcpp'
>>> 
```
Looking into my virtual environment site-packages directory `venv/lib/python3.10/site-packages` I see the following files pertaining to zxing-cpp:
```sh
zxing_cpp-2.2.0.dist-info
zxingcpp.cpython-312-darwin.so
```
Hence it looks like that during installation zxing-cpp for python 3.12 was installed instead of for python 3.10.

However, if I go through the same procedure but for Python 3.12, I can import the module in python 3.12 but running the datamatrix reader does not work.

**Comment by axxel** [comment]
Thanks for reporting back. Since I am not aware of changing anything regarding the macOS builds, I would assume that the currently published wheels on pypi.org have the same issue. Can you uninstall all your local experiments and do a simple `pip install zxing-cpp` in your Python 3.10 environment and see whether that then installs the correct `zxingcpp.cpython-310-darwin.so` file?

**Comment by kimstp** [comment]
I uninstalled zxing-cpp from my virtual environment and installed the binary wheel for python 3.10:
```sh
(venv) kimstp@hodor NHMDlabelreader % pip install zxing-cpp
Collecting zxing-cpp
  Using cached zxing_cpp-2.2.0-cp310-cp310-macosx_10_9_universal2.whl.metadata (1.9 kB)
Using cached zxing_cpp-2.2.0-cp310-cp310-macosx_10_9_universal2.whl (1.5 MB)
Installing collected packages: zxing-cpp
Successfully installed zxing-cpp-2.2.0
```
and the following got installed:
```sh
(venv) kimstp@hodor NHMDlabelreader % ls venv/lib/python3.10/site-packages/zxing* 
venv/lib/python3.10/site-packages/zxingcpp.cpython-310-darwin.so

venv/lib/python3.10/site-packages/zxing_cpp-2.2.0.dist-info:
INSTALLER	RECORD		WHEEL
METADATA	REQUESTED	top_level.txt
```
So the .so  file is correct for python 3.10.

**Comment by axxel** [comment]
As you can see from the log above, it took me ages to track this down and find a solution. :-(. Anyway, I believe I definitively fixed the problem with packaging the wrong python version build. And I might have fixed the issue with the missing c++20 support for the macOS python module as well. If not, I know what to do next.

@kimstp could you please download [this file](https://github.com/axxel/zxing-cpp/actions/runs/7982544934/artifacts/1261465928) and see whether the contained binary wheel is now a) installable b) works after installation and c) detects your datamatrix sample that needs c++20 to work?

**Comment by kimstp** [comment]
So you managed to fix a) and b), but unfortunately not c).
I tested on Python 3.9, 3.10, 3.11, and 3.12 by successfully installing the matching package from the new zip file. Then I tested with my test image from my Jan 4 post:
```python 
>>> import cv2
>>> import zxingcpp
>>> img = cv2.imread("NHMD-681096.jpg")
>>> dmcodes = zxingcpp.read_barcodes(img, formats=zxingcpp.DataMatrix)
>>> print(dmcodes)
[]
```
No detection, but the rest is working, so it looks like there is still a c++20 issue.

**Comment by axxel** [comment]
Thanks for checking. The c++20 issue is apparently a matter of setting the right value for `MACOSX_DEPLOYMENT_TARGET`, see [here](https://cibuildwheel.readthedocs.io/en/stable/cpp_standards/#macos-and-deployment-target-versions). I came to the conclusion that 10.13 has a high chance of working. The goal should obviously be to set this number as low as possible while fixing the issue to support as many macOS installations as possible. I could now simply try 10.14 next but that could end up being a boring trial and error approach ;).

Do you have an opinion about which macOS version is the oldest that is still very much relevant?

**Comment by kimstp** [comment]
I would suggest you try macos 10.15 (Catalina) - then you  would be able support hardware dating back to 2012.
Next one is macos 11 (Big Sur) (Apple changed version numbering system). The current version is macos 14 (Sonoma).

Don't know which ones would work the best, but with 10.15 you have a long hardware coverage backwards.

**Comment by kimstp** [comment]
Just an addition: If you know which C++20 features you are missing you could look them up in [this table](https://en.cppreference.com/w/cpp/compiler_support#cpp20) and see what minimum version of Apple clang is required and cross reference with this [list](https://developer.apple.com/xcode/cpp/).

**Comment by kimstp** [comment]
Ok, I pressed "comment" a bit too fast. What I wanted to write was that you can use this lists to figure out a version of Xcode that supports the needed c++20 features and the look up in [this list](https://developer.apple.com/support/xcode/)what minimum macos version is required. Hope it makes sense.

**Comment by axxel** [comment]
Thanks for the links. I did exactly that before and concluded that 10.13 has a high chance of working. :-/. The main feature required is coroutines. And that has "partial" support since XCode 10.0. Anyway, let's hope 10.15 is sufficient. If you would be so kind and test that again? Just one python version is sufficient. wheels are [here](https://github.com/zxing-cpp/zxing-cpp/actions/runs/8008182865/artifacts/1267208971).

If that does not work, then maybe a way more efficient way of finding out what is required would be if you executed
```sh
MACOSX_DEPLOYMENT_TARGET="10.15" python -m pip install zxing-cpp==2.2.0 --no-binary zxing-cpp
```
on your machine (after making sure all earlier installations are gone). And then try different version numbers. There must be one that works, since I understood that this line works on your machine with the default setting for `MACOSX_DEPLOYMENT_TARGET`.

**Comment by kimstp** [comment]
@axxel I tried to install a wheel from the new zip file, but the detection of a data matrix still does not work.

As suggested I then tried to build for a specific macos target, but I seem to have some problem with this:
```sh
(venv) kimstp@hodor NHMDlabelreader % MACOSX_DEPLOYMENT_TARGET="10.15" python -m pip install zxing-cpp==2.2.0 --no-binary zxing-cpp --no-cache-dir
Collecting zxing-cpp==2.2.0
  Downloading zxing-cpp-2.2.0.tar.gz (652 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 652.8/652.8 kB 12.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: zxing-cpp
  Building wheel for zxing-cpp (pyproject.toml) ... done
  Created wheel for zxing-cpp: filename=zxing_cpp-2.2.0-cp310-cp310-macosx_13_0_x86_64.whl size=792551 sha256=ab4be6e1958b92af3598f3f16b9ec6fa045a30d6e7488bf515cabec3eca7c32e
  Stored in directory: /private/var/folders/0m/yfsc5z1x50z9dtxgvd7bjj280000gp/T/pip-ephem-wheel-cache-boexw23b/wheels/af/a7/81/4a1ce4cf34b65c990a6199e8d573045e8baa9daa13d7a29d62
Successfully built zxing-cpp
Installing collected packages: zxing-cpp
Successfully installed zxing-cpp-2.2.0
```
As can be seen, it appears like pip ignores the enviroment variable and build towards my version of macos 13.
I will try to look a bit more into this and see if I can get the deployment target to work on my machine, but it will take me some time.

**Comment by axxel** [comment]
> As can be seen, it appears like pip ignores the enviroment variable and build towards my version of macos 13.

That conclusion might be wrong. As far as I understand, this environment variable is read by Apple's llvm. How pip decides to name a file might not have anything to do with it. You can try if it works by setting it to "10.9", which should then make your own build cause a detection failure. If that is not the case, then you are either right about pip overwriting this setting or that does not have anything to do with the issue at all and I'm at a loss a.t.m.

**Comment by kimstp** [comment]
I think my problem is that my python is build for macos 13 and according to [this](https://stackoverflow.com/questions/33184316/os-x-python-can-i-explicitly-set-macosx-deployment-target-for-extensions) I can only change deployment targets to later versions. 

I have an old mac mini that runs macos 10.15, and I am currently setting it up so I can test and build on that. I'll give you a status later.

**Comment by kimstp** [comment]
@axxel I just tried to build for target 10.9 on my macos 13 machine and it seems that it build for target 13, since the data matrix detection works:
```sh
(venv) kimstp@hodor NHMDlabelreader % MACOSX_DEPLOYMENT_TARGET="10.9" python -m pip install zxing-cpp==2.2.0 --no-binary zxing-cpp --no-cache-dir
Collecting zxing-cpp==2.2.0
  Downloading zxing-cpp-2.2.0.tar.gz (652 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 652.8/652.8 kB 12.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: zxing-cpp
  Building wheel for zxing-cpp (pyproject.toml) ... done
  Created wheel for zxing-cpp: filename=zxing_cpp-2.2.0-cp310-cp310-macosx_13_0_x86_64.whl size=792498 sha256=df595b0cf5d6ca8d8717628aa75cda1c42a54b5ca34e41d41140aacf772cf034
  Stored in directory: /private/var/folders/0m/yfsc5z1x50z9dtxgvd7bjj280000gp/T/pip-ephem-wheel-cache-h209vvwt/wheels/af/a7/81/4a1ce4cf34b65c990a6199e8d573045e8baa9daa13d7a29d62
Successfully built zxing-cpp
Installing collected packages: zxing-cpp
Successfully installed zxing-cpp-2.2.0

(venv) kimstp@hodor NHMDlabelreader % python testzxing.py 
[<zxingcpp.Result object at 0x109b263b0>]
NHMD-681096
```
I am now certain that you can only build using pip for targets above the target of your python installation, since I experimented with choosing a higher target 14 and it seems to change the file name appropriately. But since, a package build to a specific macos target will only work on a machine with the OS version equal to or higher, I have a hard time testing this. The Xcode version sets a limit on what target version you can build for.

I also now tried to build on my macos 10.15 machine and building for target 10.15, leads to an installation where the data matrix detection does not work (as expected). That machine has Xcode version 12.4 (12D4e) installed with 
```sh
kimstp@Kims-Mac-mini ~ % clang --version
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

To sum up - we know that the data matrix detection does not work when build on and for macOS 10.15 and below, and that it works when build on and for macOS 13 with Xcode 15.2.

**Comment by axxel** [comment]
Interesting but there is one thing I don't understand:

> Then I tried to take the wheel I build for target 10.9 on macOS 13

I thought you were certain that this is not possible, that `pip` will only build for targets 13.0 and up?

What you found kind of sound like the value of `MACOSX_DEPLOYMENT_TARGET` has not effect on the question of whether the detection ends up working or not *and* it seems to have no effect on whether the binary is then actually running on an older system or not? That is indeed not even super surprising, because the problems that might potentially arise stem from missing standard library support, like e.g. missing `std::filesystem` but the coroutine stuff might be a purely language/compiler property.

With all that said, the question how to make cibuildwheel do what we want is still open. The version on which the building itself happened until now is macos-latest (meaning [macos-12](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories)). Given that you successfully built it on macos-13 and that is also available as a separate runner in GHA, I switched over to that. I also sneaked in a compilation error in the c++17 code path. Since this [compiled](https://github.com/axxel/zxing-cpp/actions/runs/8031135484/job/21939163941) on macos now, the c++20 code was definitively used. Interestingly the wheel names (see [this artifact](https://github.com/axxel/zxing-cpp/actions/runs/8031135484/artifacts/1272030021) contain the string "10_15" according to the set deployment target.

This made me update the script again and remove the explicit `MACOSX_DEPLOYMENT_TARGET`, which means it defaults back to 10.9 and tried which also [works](https://github.com/axxel/zxing-cpp/actions/runs/8031636551/job/21940164298). So, if you could quickly confirm that those [wheels](https://github.com/axxel/zxing-cpp/actions/runs/8031636551/artifacts/1272100843) now properly work on both of your systems, we could finally put this to rest. sigh...

**Comment by kimstp** [comment]
@axxel I realise that the reply I wrote yesterday seems a bit contradicting. The problem is that the build process produces wheel files with names that are not consistent. I don't understand why pip behaves like this.

Anyway, the new wheels you just send works on my macos 13 machine - the installation works and detecting a data matrix works. But on my macos 10.15 machine the installation works, but I get a segmentation fault when I try to run my test script.
I think the problem is that in the new builds you are using universal2 which as far as I know is not supported by 10.15 (it was introduced in macos 11). You may have to produce separate wheels for Intel and Apple CPUs instead of relying on the universal build (packaging versions for both types of CPUs). Alternatively, you could decide to only provide builds for macos 11 and onward and then your current build process works.

I only have intel CPU machines so I will not be able to test if it works on a Apple CPU.

**Comment by axxel** [comment]
Thank you very much for all your time spent on debugging this. While it is unfortunate that the universal build does not work on 10.15 and it is definitively possible to build an extra version for Intel only, this has been the case for 2.2.0 already and until now, no-one complained about this segfault. I conclude that the the python lib is not used on any of those old systems. I can either add this extra support later if there is real demand or people can use the `build from source` approach.

**Comment by axxel** [comment]
Fixed in https://github.com/zxing-cpp/zxing-cpp/commit/c0b7f4923c4ad1abb3b28507e44e0168c3d666ee.
